pil-analyzer/src/expressionizer.rs

```diff
@@ -0,0 +1,403 @@
+//! Component that turns data from the PILAnalyzer into Analyzed,
+//! i.e. it turns more complex expressions in identities to simpler expressions.
+
+use std::collections::BTreeMap;
+use std::ops::Bound;
+
+use num_traits::sign::Signed;
+
+use powdr_ast::analyzed::{
+    self, AlgebraicBinaryOperation, AlgebraicExpression, AlgebraicReference,
+    AlgebraicUnaryOperation, Challenge, Expression, FunctionValueDefinition, PolynomialReference,
+    PolynomialType, Reference,
+};
+use powdr_ast::parsed::{
+    self,
+    types::Type,
+    visitor::{AllChildren, ExpressionVisitable},
+    ArrayLiteral, BinaryOperation, BlockExpression, FunctionCall, FunctionKind, IndexAccess,
+    LambdaExpression, LetStatementInsideBlock, Number, Pattern, TypedExpression, UnaryOperation,
+};
+use powdr_number::{BigUint, FieldElement};
+use powdr_parser_util::SourceRef;
+
+use crate::evaluator::{self, Closure, EvalError, Value};
+
+/// Turns a runtime value (usually a closure) into a FunctionValueDefinition
+/// (i.e. an expression) and sets the expected function kind.
+/// Does allow some forms of captured variables by prefixing them
+/// via let statements.
+pub fn try_to_function_value_definition<T: FieldElement>(
+    poly_id_to_name: &BTreeMap<(PolynomialType, u64), String>,
+    value: &Value<'_, T>,
+    expected_kind: FunctionKind,
+) -> Result<FunctionValueDefinition, EvalError> {
+    let mut e = Expressionizer { poly_id_to_name }.try_value_to_expression(value)?;
+
+    // Set the lambda kind since this is used to detect hints in some cases.
+    // Can probably be removed once we have prover functions.
+    if let Expression::LambdaExpression(_, LambdaExpression { kind, .. }) = &mut e {
+        if *kind != FunctionKind::Pure && *kind != expected_kind {
+            return Err(EvalError::TypeError(format!(
+                "Expected {expected_kind} lambda expression but got {kind}.",
+            )));
+        }
+        *kind = expected_kind;
+    }
+
+    Ok(FunctionValueDefinition::Expression(TypedExpression {
+        e,
+        type_scheme: None,
+    }))
+}
+
+/// Tries to convert an evaluator value to an expression with the same value.
+pub fn try_value_to_expression<T: FieldElement>(
+    poly_id_to_name: &BTreeMap<(PolynomialType, u64), String>,
+    value: &Value<'_, T>,
+) -> Result<Expression, EvalError> {
+    Expressionizer { poly_id_to_name }.try_value_to_expression(value)
+}
+
+struct Expressionizer<'a> {
+    poly_id_to_name: &'a BTreeMap<(PolynomialType, u64), String>,
```

Comment:
`(PolynomialType, u64)` is essentially `PolyID`, except that it stores the type first, which makes the `range` thing work, right? In that case, could we just change `PolyID` to store the type first?

But on a higher-level note, why don't we have something like `array_elements: BTreeMap<PolyID, (String, usize)>` (if a polynomial is an array element, maps it to the array name and index) instead? That would get rid of the `poly_id_to_name.range(...).last()` thing below, which seems kinda hacky...

Reply:
Especially since, in my understanding, the `array_elements` map would not change during the condenser run, would it? I feel like this could simplify some code further.

Reply:
I thought about changing `PolyID` but decided against because it's too fragile. We would need a static assert here to prevent other changes to `PolyID` from breaking the assumptions here.

I don't want to store an element for every single array element because I wanted to avoid performance problems coming from using large arrays.

The condenser run can actually add new arrays any time and we need to be able to handle them.

Reply:
> I thought about changing `PolyID` but decided against because it's too fragile. We would need a static assert here to prevent other changes to `PolyID` from breaking the assumptions here.

Agreed.

> The condenser run can actually add new arrays any time and we need to be able to handle them.

Right, I didn't see that `Condenser::new_column` (the only place that changes `Condenser::poly_id_to_name`) can actually add arrays of columns as well.

> I don't want to store an element for every single array element because I wanted to avoid performance problems coming from using large arrays.

We are only talking about arrays of columns, aren't we? So at most, they will be in the 1000s, because otherwise we'd have other problems, right?